DDF-2008 Merge registry report action provider from ddf-registry into ddf

mcalcote opened the pull request:
#### What does this PR do?
Create ActionProvider for registry metacards to provide an action (url). Currently this ActionProvider link is accessible when searching for metacards with the registry tag in the search UI and then checking the available actions. 

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@rzwiefel @kcwire @coyotesqrl 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@pklinef
@shaundmorris
#### How should this be tested?
Build, Confirm ActionProvider link in Search UI

This can be tested by following these steps:
1. Installing the registry app
2. Ingesting a registry metacard `ingest -t rim:RegistryPackage /Users/PathToYourDDF/ddf/catalog/spatial/registry/registry-ebrim-transformer/src/test/resources/csw-full-registry-package.xml`
3. Ingesting a second non-registry metacard
4. Open the search UI and do a wildcard search, the use the Details option to search for a **metacard-tags** contains **registry** search
5. Select the registry metacard returned by this second search and select the Action tab
6. Observe the URL generated when selecting **Export Registry Metacard Information** -> Display in new tab/window

#### Any background context you want to provide?
The code to generate the resulting report at the link provided is in another module. 
#### What are the relevant tickets?
DDF-2008
DDF-2059
#### Screenshots
![ddf2008_1](https://cloud.githubusercontent.com/assets/11355332/15477688/de3f7ed6-20ca-11e6-8d9a-bad259b7f6db.png)
![ddf2008_2](https://cloud.githubusercontent.com/assets/11355332/15477694/e2d44170-20ca-11e6-8502-1d5f03aecd14.png)
#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

shaundmorris commented on catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,147 @@
+                metacardId = URLEncoder.encode(metacard.getId(), (StandardCharsets.UTF_8).toString());
+                sourceId = URLEncoder.encode(getSource(metacard), (StandardCharsets.UTF_8).toString());
+            } catch (UnsupportedEncodingException e) {
+                LOGGER.info("Unsupported Encoding exception", e);
```
I would recommend this log level at debug since at info the stacktrace could be pretty noisy. Was there something that happens here that you think would be useful at info level?

coyotesqrl replied:
I don't see why you'd bother logging it at all, honestly. It's just not going to happen. Maybe WARN or ERROR level just to CYA?

mcalcote replied:
Yeah, it shouldn't ever happen, which is why I left it INFO, but that level doesn't really capture the essence of the message. Changed to warn.

shaundmorris commented on catalog/spatial/registry/registry-report-action-provider/src/test/java/org/codice/ddf/registry/report/action/provider/TestRegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,397 @@
+    @Test
+    public void testUriSyntaxException() {
+
+        MetacardImpl metacard = new MetacardImpl();
```
You should consider something like adding `private Metacard getSampleMetacard()`. Basically you could create a `MetacardImpl` and set the sample ID and/or registry tag in that method. It would reduce the duplication as most of these test methods start with the same logic for building a sample Metacard.

mcalcote replied:
Refactored a bit - Shaved 100+ lines off the file 👍

shaundmorris commented on catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,147 @@
+            Metacard metacard = (Metacard) subject;
+
+            if (StringUtils.isBlank(metacard.getId())) {
+                LOGGER.info("No id given. No action to provide.");
```
Recommend this at debug level since we wouldn't want this to show up for every metacard that doesn't have the registry tag.

mcalcote replied:
Good point. Fixed.

shaundmorris commented on catalog/spatial/registry/registry-report-action-provider/src/test/java/org/codice/ddf/registry/report/action/provider/TestRegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,397 @@
+
+        metacard.setId(SAMPLE_ID);
+
+        RegistryMetacardActionProvider actionProvider = new RegistryMetacardActionProvider(
```
Did you consider making the actionProvider a class variable and re-using it? I tried it out. Removed about 28 lines and the unit tests still passed. Seems like you don't really need to change the configuration on it on each test.

mcalcote replied:
Good Call 👍

coyotesqrl commented on catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,147 @@
+    private Action getAction(String metacardId, String sourceId) {
+
+        if(StringUtils.isNotBlank(sourceId)) {
+            sourceId = SOURCE_ID_QUERY_PARAM + sourceId;
```
I don't like to see parameters modified like this on the stack. I'd much rather see a new String created and used throughout the method.

shaundmorris replied:
+1. We shouldn't re re-assigning the parameter. I'm surprised findbugs didn't have a rule violation for this.

mcalcote replied:
Fixed, will avoid in the future.

coyotesqrl commented on catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,147 @@
+        URL url;
+        try {
+
+            URI uri = new URI(SystemBaseUrl.constructUrl(REGISTRY_PATH + "/" + metacardId + REPORT_PATH + FORMAT + sourceId,
```
Instead of all this String concatenation...
```java
new URI(SystemBaseUrl.constructUrl(String.format("%s/%s%s%s%s", REGISTRY_PATH, "/", metacardId, REPORT_PATH, FORMAT, sourceId)));
```

mcalcote replied:
Replaced concat

coyotesqrl commented on catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,147 @@
+            url = uri.toURL();
+
+        } catch (MalformedURLException e) {
+            LOGGER.info("Malformed URL exception", e);
```
DEBUG

coyotesqrl commented on catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,147 @@
+            LOGGER.info("Malformed URL exception", e);
+            return null;
+        } catch (URISyntaxException e) {
+            LOGGER.info("URI Syntax exception", e);
```
DEBUG

coyotesqrl commented on catalog/spatial/registry/registry-report-action-provider/src/main/java/org/codice/ddf/registry/report/action/provider/RegistryMetacardActionProvider.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    private boolean isHostUnset(String host) {
+        return (host == null || host.trim()
```
Would an empty string also be considered unset?

mcalcote replied:
I think `SystemBaseUrl` has default string values when `System.setProperty` isn't called, and if someone set the property to an empty string, there would be other exceptions/failures before we got here.

coyotesqrl commented:
Giving this a 👎 for now. Needs a sample registry metacard, a screenshot, a look over the heroing instructions, and some work on the feature in order to be good to go.

vinamartin commented:
@coyotesqrl A sample registry metacard won't be necessary for this hero. 
To test this, first, start up the registry app. Then ingest any metacards you like, it doesn't matter what they are. After you've done that, go into the search UI and do a query for metacards with tags that contain the word `registry`. You should see a metacard which represents the current DDF. Verify that its actions include `Export Registry Metacard Information` and that clicking it goes to the desire URL. The actual report is coming to a pull request near you.

coyotesqrl commented:
Thanks for that clarification, @vinamartin. I'm still holding on approving for now until the features are updated but it'll be easier to test now that I know I don't need a special metacard to ingest at all.

mcalcote commented:
I updated the Testing Instructions and added screenshots to the PR description. Please follow those steps to test the registry metacard action provider.

coyotesqrl commented:
I rescind my thumbs down. 

Heroed successfully after updates. 👍

shaundmorris commented:
👍  verified test steps.
